ISLANDORA-2175: Use appropriate namespace in CSV importer

d-r-p opened the pull request:
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2175

# What does this Pull Request do?

Currently, the Entities CSV importer creates all new objects with namespace 'islandora'. Through this PR, the importer assigns to the new objects the namespace specified in the policy of the collection the user wants to ingest into.

The code is based on an old (2016-09-27) PM from @nhart - thank you!

Additionally, this PR prevents the ingestion of orphaned objects (i.e. those which do not belong to any collection), in case the user chooses a non-existent collection PID. In that case, the new objects are now created in the root collection.

# What's new?

* Get the namespace for the new objects from the collection policy rather than
    imposing 'islandora'
* Default to ingesting into islandora:root if the user-provided collection PID
    is invalid

Note: the second behaviour could easily be removed by deleting [these five lines](https://github.com/d-r-p/islandora_solution_pack_entities/blob/540ae74cd38c45e5c5b9afb17131a32597198f6b/modules/islandora_entities_csv_import/includes/utilities.inc#L88-L92).

# How should this be tested?

E.g., inside a VM:
* If not already present, create a collection with collection policy 'islandora:personCModel' enabled and a different namespace than 'islandora'
* Upload an appropriate (see [here](https://wiki.duraspace.org/display/ISLANDORA/Entities+Solution+Pack)) CSV into this collection (on /import_entity_csv) and convince yourself that all newly created objects have namespace 'islandora'
* _For the second feature only_: upload a CSV into a non-existent collection (write garbage in the PID field of the form) and verify that the newly created objects are not members of any collection
* Merge the PR
* Re-upload the CSV and verify that the newly created objects have the namespace you set up in the collection policy management
* _For the second feature only_: upload a CSV into a non-existent collection and verify that the newly created objects are members of the root collection

# Additional Notes:

* Does this change require documentation to be updated? **no**
* Does this change add any new dependencies? **no**
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? **no**
* Could this change impact execution of existing code? **no**

# Interested parties
@rosiel @Islandora/7-x-1-x-committers

d-r-p committed:
ISLANDORA-2175: Use appropriate namespace in CSV importer

- Get the namespace for the new objects from the collection policy rather than
  imposing 'islandora'
- Default to ingesting into islandora:root if the user-provided collection PID
  is invalid

willtp87 commented on modules/islandora_entities_csv_import/includes/utilities.inc:
```diff
@@ -68,10 +68,31 @@ function islandora_entities_read_csv($path) {
+  elseif (!$collection_object) {
+    // If the user-provided $collection_pid is invalid, we ingest into
+    // islandora:root (otherwise the object might be out of sight).
+    $collection_pid = 'islandora:root';
```
I think this would be rather unexpected.  It could perhaps be controlled by some sort of flag.

d-r-p commented:
I absolutely do not insist on this functionality as I am aware that it is not ideal (see also my comment in the [JIRA ticket](https://jira.duraspace.org/browse/ISLANDORA-2175?focusedCommentId=58438&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-58438)). The thought behind it was that it can happen quite easily to not properly autocomplete the collection PID in the upload form. This would result in the creation of (potentially) a lot of out-of-sight objects. If they were in the root collection, a few clicks could migrate them into the appropriate collection (albeit with the wrong namespace) or delete them. But, by all means, delete [those five lines](https://github.com/d-r-p/islandora_solution_pack_entities/blob/540ae74cd38c45e5c5b9afb17131a32597198f6b/modules/islandora_entities_csv_import/includes/utilities.inc#L88-L92).

DiegoPino commented:
@d-r-p I would say, if the parent collection does not exist, then don't ingest (preprocess should actually fail), could be easily added as part of a pre validation. But not sure if that is part of this JIRA issue. Thanks for this, all this cleanup is very appreciated. Do you think you have time to offer to this Release Process as tester, documenter, Module Maintainer or bug fixer? Let me know. Best!

d-r-p committed:
Implement changes suggested by @willtp87 and @DiegoPino

- Remove default behaviour of ingesting into `islandora:root`
  and just warn instead
- Implement a validator for the form element to prevent the
  issue from arising in the first place

rosiel commented:
@d-r-p ++ Haven't run it, but it looks good (and useful) to me!

d-r-p commented:
Pinging @Islandora/7-x-1-x-committers: anyone happy to review this before the 7.x-1.11 code freeze?

DiegoPino commented:
@d-r-p will have this done before code freeze, away on a conference and with little time, but will for sure on Friday, 24 hours and then merge if all good. But if not all good: this sounds like a bugfix, and bugfixes can get into the release after code freeze.

d-r-p commented:
Thank you, @DiegoPino!

DiegoPino commented on modules/islandora_entities_csv_import/includes/islandora_entities_csv_import.form.inc:
```diff
@@ -57,3 +58,42 @@ function islandora_entities_import_form_submit(array $form, array &$form_state)
+ *   The Drupal Form State.
+ */
+function islandora_entities_csv_import_validate_collection(array $element, array &$form_state) {
+  $collection_pid = &$form_state['values']['collection'];
```
@d-r-p just curiosity. Why are you loading the collection_pid by reference here?

d-r-p replied:
Bad copy&paste, I guess... I will remove it when cleaning up.

DiegoPino commented on modules/islandora_entities_csv_import/includes/utilities.inc:
```diff
@@ -68,10 +68,26 @@ function islandora_entities_read_csv($path) {
   $repository = $tuque->repository;
-  $fedora_object = $repository->constructObject('islandora');
+  $fedora_object = $repository->constructObject($namespace);
   $fedora_object->label = isset($scholar['DISPLAY_NAME']) ? $scholar['DISPLAY_NAME'] : 'Member';
```
Just in case we missed some checks, i would make sure we trim and then we check if not empty (v/s) not set in $scholar['DISPLAY_NAME']

d-r-p replied:
I am not sure, I understand well, but an additional check is probably a good idea...

DiegoPino commented on modules/islandora_entities_csv_import/includes/utilities.inc:
```diff
@@ -68,10 +68,26 @@ function islandora_entities_read_csv($path) {
-  $fedora_object = $repository->constructObject('islandora');
+  $fedora_object = $repository->constructObject($namespace);
   $fedora_object->label = isset($scholar['DISPLAY_NAME']) ? $scholar['DISPLAY_NAME'] : 'Member';
   $fedora_object->owner = 'admin';
```
Are we sure Admin is the owner here? I feel this should be the actual global $user;

d-r-p replied:
Probably not. This is the old code, though...

DiegoPino commented on modules/islandora_entities_csv_import/includes/utilities.inc:
```diff
@@ -82,6 +98,13 @@ function islandora_entities_build_scholar_object($scholar, $collection_pid) {
   $fedora_object->relationships->add(FEDORA_MODEL_URI, 'hasModel', 'islandora:personCModel');
   $fedora_object->relationships->add(FEDORA_RELS_EXT_URI, 'isMemberOfCollection', $collection_pid);
   $new_fedora_object = islandora_add_object($fedora_object);
+  if ($new_fedora_object && !$collection_object) {
```
I'm confused here. Why are we allowing orphan objects to be ingested?

d-r-p replied:
I thought this was the expected legacy behaviour (but we now issue a warning). It should not happen when coming from the form, but if someone were to call the function directly, we keep doing what we always did. I am happy to remove this, though...

DiegoPino commented on modules/islandora_entities_csv_import/includes/utilities.inc:
```diff
@@ -68,10 +68,26 @@ function islandora_entities_read_csv($path) {
+  // Load the collection object and, upon success, get the appropriate
+  // namespace.
+  $collection_object = islandora_object_load($collection_pid);
+  if ($collection_object && isset($collection_object['COLLECTION_POLICY'])) {
```
should we not simply drop ingest if the parent collection does not exist?

DiegoPino commented on modules/islandora_entities_csv_import/includes/islandora_entities_csv_import.form.inc:
```diff
@@ -57,3 +58,42 @@ function islandora_entities_import_form_submit(array $form, array &$form_state)
+function islandora_entities_csv_import_validate_collection(array $element, array &$form_state) {
+  $collection_pid = &$form_state['values']['collection'];
+  $collection_object = islandora_object_load($collection_pid);
+  if ($collection_object) {
```
From line 75, most of the code looks pretty similar to the one in each individual batch process. Seems to me we could make a single "check if collection exists and policy matches what we are passing" function and reuse it here and in batch handler down here https://github.com/Islandora/islandora_solution_pack_entities/pull/125/files#diff-100a7d9d528fcbde5b9c60729bd2d173R76? ?

d-r-p replied:
You are right, one could pull out some re-used code into a custom function...

DiegoPino commented on modules/islandora_entities_csv_import/includes/utilities.inc:
```diff
@@ -68,10 +68,26 @@ function islandora_entities_read_csv($path) {
+  $namespace = 'islandora';
+  // Load the collection object and, upon success, get the appropriate
+  // namespace.
+  $collection_object = islandora_object_load($collection_pid);
```
Ok, i feel we need to refactor something here. We are passing $collection_pid as a single value for the whole batch processor, not per object, right? That means that checking for each object over and over is an overkill in terms of processing (more if we can drop batches for non existing parent objects completely). I would move the whole "is the collection there and does it has a valid policy) out of islandora_entities_build_scholar_object and into whatever triggers the batch. Right now it seems to be only the form, but it could be also a drush script right? Maybe there is a need i did not understood here, but open to ideas and understanding this

d-r-p replied:
You are probably right. But this would mean to change the API of `islandora_entities_build_scholar_object()` to hand over the namespace as well, right? Is this wanted?

DiegoPino reviewed (changes requested):
Hi @d-r-p, I have a few concerns. Let me know what you think and if you believe those are valid. Thanks! We still have 48 hours to get this into the release.

DiegoPino commented:
@d-r-p you are right. Most of the issues here are just legacy code (sorry 😢 ), so my inclination here would be, for the sake of advancing instead of getting stuck(me making it getting stuck) on fixing all back issues,  to fix  internals and pre invocation of `islandora_entities_build_scholar_object($scholar, $collection_pid)`  and avoid checking collection existence and namespace for every object ingest in the batch. I find it an overkill really. Means, yes, adding collection checks before the individual operations, could be at https://github.com/Islandora/islandora_solution_pack_entities/blob/704be6507012698a17d85bd3fb2c0579fd96e35e/modules/islandora_entities_csv_import/includes/utilities.inc#L16 or somewhere before in the submit handler.  (which means also unifying the checks in a single function, validation and submit makes sense), but this can be archived without changing the function signature (API mention) by passing needed info inside the $scholar array. @bryjbrown @bondjimbond @rosiel what do you think about this? 

I can be pretty picky 😆  and know I have to acknowledge that most of my concerns come from old code (but so so close to the new one!!)

d-r-p commented:
@DiegoPino well, I would rather have the namespace shipped in an _optional_ argument in both `islandora_entities_build_scholar_object()` and `islandora_entities_import_csv_batch()`, or would this break something (especially for the latter)? The checks and namespace retrieval could then be performed in the submit handler of the form, and we would avoid the repeated expensive checks. By the way, from what I can tell (`grep`), neither function is called anywhere else in a standard vagrant machine...

d-r-p committed:
Implement changes requested by @DiegoPino

- Perform all collection related checks in submit handler
- Add `$namespace` as an optional parameter to
  `function islandora_entities_import_csv_batch()` and
  `islandora_entities_build_scholar_object()`
- Treat whitespace only object label as empty
- Set current user as owner of created objects

rosiel commented:
@d-r-p I like the use of an optional argument, that allows legacy compatibility as well as improved functioning. Checking only once, in the form submit, seems reasonable. The code looks good to me.

d-r-p commented:
Thank you, @rosiel! There is just one tiny thing that keeps bothering me: the check for the file is done through validation, whereas the check for the collection is done upon submission. This means that, if both form elements are invalid, only the flaw with the file will show up at first, and the flaw with the collection will show up only after the file part has been fixed. That is a bit contrary to what a user might expect, I think. There would be a few possible solutions to this:

- (simplest) the collection could be checked in a validation function as before, and we retrieve the namespace again upon submission
- we define a function that retrieves all the necessary info (setting to `NULL` or `FALSE` if erroneous), and call that twice (this would have the additional benefit for being re-usable if someone would like to make a `drush` script for entity import)
- we put the namespace in a hidden form field in the validation step (with the benefit of only running the checks once)

What do you guys think?

DiegoPino commented:
@d-r-p i agree on all your points there. Do you think you can get that done by today? If not I will approve anyway because it is much better than the original functionality anyway. Thanks a lot

d-r-p commented:
@DiegoPino @rosiel I sure can. But which alternative (or a combination) would you prefer?

DiegoPino commented:
This combo (both options are mostly the same right?), means we check on validation and on submit (submit, not on every single object) but using a single common function.  I feel its good and it matches one of my review comments too.

> - (simplest) the collection could be checked in a validation function as before, and we retrieve the namespace again upon submission
> - we define a function that retrieves all the necessary info (setting to NULL or FALSE if erroneous), and call that twice (this would have the additional benefit for being re-usable if someone would like to make a drush script for entity import)

d-r-p committed:
Improve form validation

- Define a function that retrieves the collection info
- Validate the 'collection' form element separately
- Retrieve namespace in submit handler and re-check
  collection info

DiegoPino commented:
@d-r-p this looks good for me. 24 hours for merge and right on time for code freeze. Does anyone want to give it the last run? Thanks so much for addressing all the concerns here. You did a good job.
